api/envoy/api/v2/core/protocol.proto

```diff
@@ -42,6 +42,24 @@ message Http1ProtocolOptions {
   // Envoy does not otherwise support HTTP/1.0 without a Host header.
   // This is a no-op if *accept_http_10* is not true.
   string default_host_for_http_10 = 3;
+
+  message HeaderKeyFormat {
+    enum HeaderFormat {
+      // By default, all headers are lower cased.
+      DEFAULT = 0;
+
+      // Formats the header using Train-Case: the first character and any character following a
+      // special character will be capitalized if it's an alpha character. For example,
+      // "content-type" becomes "Content-Type", and "foo$b#$ar" becomes "Foo$B#$Ar".
+      TRAIN_CASE = 1;
```

Comment:
Is "train case" a standard term? I haven't heard it before.

Reply:
i hadn't heard about it either before it was used in the issue that was filed. Googling around I can find some references to it (https://docs.rs/Inflector/0.7.0/inflector/cases/traincase/index.html, https://medium.com/infancyit/letter-case-special-styles-7fa369bf173d), though other sources have it mean something else (https://en.wikipedia.org/wiki/Letter_case#Special_case_styles says TRAIN-CASE, https://www.fluentcpp.com/2019/01/08/restmycase-a-c-library-for-formatting-string-cases/ says train_Case).

Open to naming suggestions here, given that this doesn't seem  to be a very common term

Reply:
One option might be `rfc7230_casing` since we're basically trying to present the headers the way they are cased in the HTTP/1.1 RFCs.

Reply:
that sounds good to me. @PiotrSikora any concerns about taking the name of the RFC in vain?

Reply:
`"TrainCase"` is called [`PascalCase (aka Upper Camel Case)`](https://en.wikipedia.org/wiki/Pascal_case).

We should also have the `DONT_MODIFY` case (probably a good default, and in theory it's backwards compatible, since HTTP headers are case-insensitive), because not all standard headers use `PascalCase` (e.g. `Content-MD5` and `TE`), and use transcoding to `PascalCase` or `lowercase` only when transforming headers between HTTP/1.1 and HTTP/2.

Reply:
Nevermind, PascalCase is without hyphens.

Reply:
@htuch using RFC number where isn't appropriate here, IMHO, because that would imply that the casing in defined in that RFC, whereas that's not the case (as far as I can tell).

Reply:
How about `propercase_words`? then we define a word as any sequence of alpha characters